// arkguard/lib/transformers/layout/DisableConsoleTransformer.js
"use strict";
/*
 * Copyright (c) 2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const typescript_1 = require("typescript");
const TransformPlugin_1 = require("../TransformPlugin");
const NodeUtils_1 = require("../../utils/NodeUtils");
const ArkObfuscator_1 = require("../../ArkObfuscator");
const PrinterUtils_1 = require("../../utils/PrinterUtils");
var secharmony;
(function (secharmony) {
    secharmony.transformerPlugin = {
        'name': 'disableConsolePlugin',
        'order': TransformPlugin_1.TransformerOrder.DISABLE_CONSOLE_TRANSFORMER,
        'createTransformerFactory': createDisableConsoleFactory
    };
    function createDisableConsoleFactory(option) {
        if (!option.mDisableConsole) {
            return null;
        }
        return disableConsoleFactory;
        function disableConsoleFactory(context) {
            return transformer;
            function transformer(node) {
                if (!(0, typescript_1.isSourceFile)(node) || NodeUtils_1.NodeUtils.isDeclarationFile(node)) {
                    return node;
                }
                (0, PrinterUtils_1.startSingleFileEvent)(PrinterUtils_1.EventList.REMOVE_CONSOLE, ArkObfuscator_1.performancePrinter.timeSumPrinter);
                let resultAst = visitAst(node);
                let parentNodes = (0, typescript_1.setParentRecursive)(resultAst, true);
                (0, PrinterUtils_1.endSingleFileEvent)(PrinterUtils_1.EventList.REMOVE_CONSOLE, ArkObfuscator_1.performancePrinter.timeSumPrinter);
                return parentNodes;
            }
            /**
             * delete console log print expression, only support simple format like:
             *  - console.xxx();
             *  - console['xxx']();
             * @param node
             */
            function visitAst(node) {
                const visitedAst = (0, typescript_1.visitEachChild)(node, visitAst, context);
                if (!((0, typescript_1.isSourceFile)(node) || (0, typescript_1.isBlock)(node) || (0, typescript_1.isModuleBlock)(node) || (0, typescript_1.isCaseClause)(node) || (0, typescript_1.isDefaultClause)(node))) {
                    return visitedAst;
                }
                //@ts-ignore
                const deletedStatements = deleteConsoleStatement(visitedAst.statements);
                if ((0, typescript_1.isSourceFile)(node)) {
                    return typescript_1.factory.updateSourceFile(node, deletedStatements);
                }
                else if ((0, typescript_1.isBlock)(node)) {
                    return typescript_1.factory.createBlock(deletedStatements, true);
                }
                else if ((0, typescript_1.isModuleBlock)(node)) {
                    return typescript_1.factory.createModuleBlock(deletedStatements);
                }
                else if ((0, typescript_1.isCaseClause)(node)) {
                    return typescript_1.factory.createCaseClause(node.expression, deletedStatements);
                }
                else {
                    return typescript_1.factory.createDefaultClause(deletedStatements);
                }
            }
            function deleteConsoleStatement(statements) {
                const reservedStatements = [];
                statements.forEach((child) => {
                    if (!isSimpleConsoleStatement(child)) {
                        reservedStatements.push(child);
                    }
                });
                return reservedStatements;
            }
            function isSimpleConsoleStatement(node) {
                if (!(0, typescript_1.isExpressionStatement)(node)) {
                    return false;
                }
                if (!node.expression || !(0, typescript_1.isCallExpression)(node.expression)) {
                    return false;
                }
                const expressionCalled = node.expression.expression;
                if (!expressionCalled) {
                    return false;
                }
                if ((0, typescript_1.isPropertyAccessExpression)(expressionCalled) && expressionCalled.expression) {
                    if ((0, typescript_1.isIdentifier)(expressionCalled.expression) && expressionCalled.expression.text === 'console') {
                        return true;
                    }
                }
                if ((0, typescript_1.isElementAccessExpression)(expressionCalled) && expressionCalled.expression) {
                    if ((0, typescript_1.isIdentifier)(expressionCalled.expression) && expressionCalled.expression.text === 'console') {
                        return true;
                    }
                }
                return false;
            }
        }
    }
    secharmony.createDisableConsoleFactory = createDisableConsoleFactory;
})(secharmony || (secharmony = {}));
module.exports = secharmony;
