# Enforce consistent spacing before and after commas. (`array-bracket-spacing`)

## Rule Details

Enforce consistent spacing before and after commas.

> See [***documentation***](https://developer.huawei.com/consumer/{{region}}/doc/harmonyos-guides-{{apiVersion}}/ide_array-bracket-spacing-{{apiVersion}}) for more details.

## Examples

Examples of **incorrect** code for this rule.

```ts
// There should be no space after '['.
// There should be no space before ']'.
let arr = [ 1, 2 ];
```

Examples of **correct** code for this rule.

```ts
let arr = [1, 2];
```
