# Enforce the consistent use of single quotes. (`quotes`)

## Rule Details

Enforce the consistent use of single quotes.

> See [***documentation***](https://developer.huawei.com/consumer/{{region}}/doc/harmonyos-guides-{{apiVersion}}/ide-quotes-stylistic-{{apiVersion}}) for more details.

## Examples

Examples of **incorrect** code for this rule.

```ts
// Strings must use single quotes.
const a = "hello";
```

Examples of **correct** code for this rule.

```ts
const a = 'hello';
const b = `hello`;
```
